### PR TITLE
[Improvement](hash) some refactor of process hash table probe impl

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes
some refactor of process hash table probe impl

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/vec/exec/vset_operation_node.cpp [hunk 1]
```diff
@@ -82,16 +82,23 @@ struct HashTableBuild {
             key_getter.set_serialized_keys(hash_table_ctx.keys.data());
         }
 
+        _build_side_hash_values.resize(_rows);
+        auto keys = key_getter.get_keys(_rows);
```
Comment (HappenLee) [inline be/src/vec/exec/vset_operation_node.cpp:86]: here better use reference, avoid copy

File: be/src/vec/exec/vpartition_sort_node.cpp [hunk 2]
```diff
@@ -112,20 +112,12 @@ void VPartitionSortNode::_emplace_into_hash_table(const ColumnRawPtrs& key_colum
                 _pre_serialize_key_if_need(state, agg_method, key_columns, num_rows);
 
                 //PHHashMap
+                auto keys = state.get_keys(num_rows);
```
Comment (HappenLee) [inline be/src/vec/exec/vpartition_sort_node.cpp:115]: use reference

File: be/src/vec/exec/vaggregation_node.cpp [hunk 3]
```diff
@@ -982,21 +982,11 @@ void AggregationNode::_find_in_hash_table(AggregateDataPtr* places, ColumnRawPtr
                 AggState state(key_columns, _probe_key_sz, nullptr);
 
                 _pre_serialize_key_if_need(state, agg_method, key_columns, num_rows);
-
+                auto keys = state.get_keys(num_rows);
```
Comment (HappenLee) [inline be/src/vec/exec/vaggregation_node.cpp:985]: same

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 4]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
+            if (non_sub_blocks_matched_row_count > 0) {
+                filter_map[end_row_idx - 1] = !filter_map[end_row_idx - 1];
+            }
         }
 
-        return Status::OK();
+        output_block->get_by_position(result_column_id).column = std::move(new_filter_column);
+    } else if constexpr (JoinOpType == TJoinOp::RIGHT_SEMI_JOIN ||
+                         JoinOpType == TJoinOp::RIGHT_ANTI_JOIN) {
+        for (int i = 0; i < column->size(); ++i) {
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:777]: add a todo，simd the code

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 5]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
+    _right_col_idx = _join_context->_is_right_semi_anti && !with_other_join_conjuncts
+                             ? 0
+                             : _join_context->_left_table_data_types->size();
+    _right_col_len = _join_context->_right_table_data_types->size();
+    _row_count_from_last_probe = 0;
+
+    _probe_indexs.clear();
+    _build_block_rows.clear();
+    _build_block_offsets.clear();
+    _visited_map.clear();
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:223]: only need do clear() in with other join conjuncts

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 6]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
     size_t probe_size = 0;
-    auto& probe_row_match_iter =
-            std::get<ForwardIterator<Mapped>>(*_join_context->_probe_row_match_iter);
+    auto& probe_row_match_iter = _probe_row_match<Mapped, with_other_conjuncts>(
+            current_offset, probe_index, probe_size, all_match_one);
+
+    bool is_the_last_sub_block = false;
+    if (with_other_conjuncts && probe_size != 0) {
+        is_the_last_sub_block = !probe_row_match_iter.ok();
+        _same_to_prev.emplace_back(false);
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:339]: if not other join conjunt do not need do the work

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 7]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
+    SCOPED_TIMER(_join_context->_process_other_join_conjunct_timer);
+    int orig_columns = output_block->columns();
+    IColumn::Filter other_conjunct_filter(row_count, 1);
+    bool can_be_filter_all = false;
+    RETURN_IF_ERROR(VExprContext::execute_conjuncts(*_join_context->_other_join_conjuncts, nullptr,
+                                                    output_block, &other_conjunct_filter,
+                                                    &can_be_filter_all));
+
+    auto result_column_id = output_block->columns();
+    auto filter_column = ColumnVector<UInt8>::create();
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:515]: better reuse the mem of filter column

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 8]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
+    if (can_be_filter_all) {
+        memset(other_conjunct_filter.data(), 0, row_count);
+    }
+    filter_column->get_data() = std::move(other_conjunct_filter);
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:519]: here do move, the can_be_filter_all is useless

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 9]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
+    output_block->insert({std::move(filter_column), std::make_shared<DataTypeUInt8>(), ""});
+    auto column = output_block->get_by_position(result_column_id).column;
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:521]: use ref

File: be/src/vec/exec/join/process_hash_table_probe_impl.h [hunk 10]
```diff
@@ -207,883 +208,601 @@ void ProcessHashTableProbe<JoinOpType>::_pre_serialize_key(
...
-Status ProcessHashTableProbe<JoinOpType>::do_process_with_other_join_conjuncts(
-        HashTableType& hash_table_ctx, ConstNullMapPtr null_map, MutableBlock& mutable_block,
-        Block* output_block, size_t probe_rows, bool is_mark_join) {
-    auto& probe_index = *_join_context->_probe_index;
-    auto& probe_raw_ptrs = *_join_context->_probe_columns;
-    if (_build_block_rows.size() < probe_rows * PROBE_SIDE_EXPLODE_RATE) {
-        _probe_indexs.resize(probe_rows * PROBE_SIDE_EXPLODE_RATE);
-        _build_block_rows.resize(probe_rows * PROBE_SIDE_EXPLODE_RATE);
-        _build_block_offsets.resize(probe_rows * PROBE_SIDE_EXPLODE_RATE);
+Status ProcessHashTableProbe<JoinOpType>::do_other_join_conjuncts(
```
Comment (HappenLee) [inline be/src/vec/exec/join/process_hash_table_probe_impl.h:497]: add comment for the param：
what means of：
1、is_the_last_sub_block
2. multi_matched_output_row_count

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/vec/exec/join/process_hash_table_probe_impl.h (hunk 10 above)
Comment (BiteTheDDDDt) [reply be/src/vec/exec/join/process_hash_table_probe_impl.h:501]: Comments are added to variable declarations

File: be/src/vec/exec/join/process_hash_table_probe_impl.h (hunk 8 above)
Comment (BiteTheDDDDt) [reply be/src/vec/exec/join/process_hash_table_probe_impl.h:521]: removed

File: be/src/vec/exec/join/process_hash_table_probe_impl.h (hunk 6 above)
Comment (BiteTheDDDDt) [reply be/src/vec/exec/join/process_hash_table_probe_impl.h:346]: This branch will be entered only when with_other_conjuncts is true

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**